contracts/utils/HashCheckpoint.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.21;
+
+/// @title Hash Checkpoint - Smart contract for checkpointing IPFS hashes on-chain
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+contract HashCheckpoint {
+    error OwnerOnly(address sender, address owner);
+    error ZeroAddress();
+    error ZeroValue();
+    event OwnerUpdated(address indexed owner);
+    event ManagerUpdated(address indexed manager);
+    event BaseURIChanged(string baseURI);
+    event HashUpdated(address indexed emitter, bytes32 hash);
+
+    // Owner address
+    address public owner;
+    // Base URI
+    string public baseURI;
+    // To better understand the CID anatomy, please refer to: https://proto.school/anatomy-of-a-cid/05
+    // CID = <multibase_encoding>multibase_encoding(<cid-version><multicodec><multihash-algorithm><multihash-length><multihash-hash>)
+    // CID prefix = <multibase_encoding>multibase_encoding(<cid-version><multicodec><multihash-algorithm><multihash-length>)
+    // to complement the multibase_encoding(<multihash-hash>)
+    // multibase_encoding = base16 = "f"
+    // cid-version = version 1 = "0x01"
+    // multicodec = dag-pb = "0x70"
+    // multihash-algorithm = sha2-256 = "0x12"
+    // multihash-length = 256 bits = "0x20"
+    string public constant CID_PREFIX = "f01701220";
+    // Map of address => latest IPFS hash
+    mapping (address => bytes32) public latestHashes;
+
+    /// @dev Hash checkpoint constructor.
+    /// @param _baseURI Hash registry base URI.
+    constructor(string memory _baseURI) {
+        baseURI = _baseURI;
+        owner = msg.sender;
+    }
+
+    /// @dev Changes the owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external virtual {
```

Comment:
Probably no need to be a virtual function? For all of them below as well.

Reply:
I guess no harm